pkg/vmprovider/providers/vsphere2/virtualmachine/backup.go

```diff
@@ -28,136 +42,174 @@ type VMDiskData struct {
 
 // BackupVirtualMachine backs up the required data of a VM into its ExtraConfig.
 // Currently, the following data is backed up:
-// - Kubernetes VirtualMachine object in YAML format (without its .status field).
-// - VM bootstrap data in JSON (if provided).
-// - VM disk data in JSON (if created and attached by PVCs).
-func BackupVirtualMachine(ctx context.BackupVirtualMachineContextA2) error {
+// - VM Kubernetes resource in YAMl.
+// - Additional VM-relevant Kubernetes resources in YAML, separated by "---".
+// - Cloud-Init instance ID (if not already stored in ExtraConfig).
+// - PVC disk data in JSON format (if DiskUUIDToPVC is not empty).
+func BackupVirtualMachine(opts BackupVirtualMachineOptions) error {
 	var moVM mo.VirtualMachine
-	if err := ctx.VcVM.Properties(ctx.VMCtx, ctx.VcVM.Reference(),
+	if err := opts.VcVM.Properties(opts.VMCtx, opts.VcVM.Reference(),
 		[]string{"config.extraConfig"}, &moVM); err != nil {
-		ctx.VMCtx.Logger.Error(err, "Failed to get VM properties for backup")
+		opts.VMCtx.Logger.Error(err, "failed to get VM properties for backup")
 		return err
 	}
-	curEcMap := util.ExtraConfigToMap(moVM.Config.ExtraConfig)
 
-	var ecToUpdate []types.BaseOptionValue
+	curEcMap := util.ExtraConfigToMap(moVM.Config.ExtraConfig)
+	ecToUpdate := []types.BaseOptionValue{}
 
-	vmKubeDataBackup, err := getDesiredVMKubeDataForBackup(ctx.VMCtx.VM, curEcMap)
+	vmYAML, err := getDesiredResourceYAMLForBackup(
+		[]client.Object{opts.VMCtx.VM},
+		vmopv1.VMResourceYAMLExtraConfigKey,
+		curEcMap,
+	)
 	if err != nil {
-		ctx.VMCtx.Logger.Error(err, "Failed to get VM kube data for backup")
+		opts.VMCtx.Logger.Error(err, "failed to get VM resource yaml for backup")
 		return err
 	}
-
-	if vmKubeDataBackup == "" {
-		ctx.VMCtx.Logger.V(4).Info("Skipping VM kube data backup as unchanged")
+	if vmYAML == "" {
+		opts.VMCtx.Logger.Info("Skipping VM resource yaml backup as unchanged")
 	} else {
 		ecToUpdate = append(ecToUpdate, &types.OptionValue{
-			Key:   vmopv1.VMBackupKubeDataExtraConfigKey,
-			Value: vmKubeDataBackup,
+			Key:   vmopv1.VMResourceYAMLExtraConfigKey,
+			Value: vmYAML,
 		})
 	}
 
-	instanceIDBackup, err := getDesiredCloudInitInstanceIDForBackup(ctx.VMCtx.VM, curEcMap)
+	additionalYAML, err := getDesiredResourceYAMLForBackup(
+		opts.AdditionalResources,
+		vmopv1.AdditionalResourcesYAMLExtraConfigKey,
+		curEcMap,
+	)
 	if err != nil {
-		ctx.VMCtx.Logger.Error(err, "Failed to get cloud-init instance ID for backup")
+		opts.VMCtx.Logger.Error(err, "failed to get additional resources yaml for backup")
 		return err
 	}
 
-	if instanceIDBackup == "" {
-		ctx.VMCtx.Logger.V(4).Info("Skipping cloud-init instance ID as already stored")
+	if additionalYAML == "" {
+		opts.VMCtx.Logger.Info("Skipping additional resources yaml backup as unchanged")
```

Comment:
These logs look like the potential for a lot of log spam in the common case for every reconcile

Reply:
Good catch. Looks like I accidentally dropped the V(4) in this log change, added it back in the latest commit.